File: frontend/lib/api/graph/alert-centric.ts
import type {
  AttackTreeResponse,
  GraphEdge,
  GraphNode,
  GraphResponse,
  GraphTreeNode,
} from "@/lib/api/graph/types"

export function isAlertNode(node: GraphNode): boolean {
  return node.group.includes("Alert")
}

/** One display edge per (from, to) for CAUSED — avoids stacked duplicate labels on the canvas. */
export function dedupeCausedEdges(edges: GraphEdge[]): GraphEdge[] {
  const byPair = new Map<string, GraphEdge>()
  const rest: GraphEdge[] = []
  for (const edge of edges) {
    if (edge.label !== "CAUSED") {
      rest.push(edge)
      continue
    }
    const key = `${edge.from}\0${edge.to}`
    const existing = byPair.get(key)
    if (!existing) {
      byPair.set(key, edge)
      continue
    }
    const prev = existing.properties ?? {}
    const next = edge.properties ?? {}
    const narrative =
      (typeof next.narrative === "string" && next.narrative) ||
      (typeof prev.narrative === "string" && prev.narrative) ||
      undefined
    byPair.set(key, {
      ...existing,
      properties: {
        ...prev,
        ...next,
        ...(narrative ? { narrative } : {}),
      },
    })
  }
  return [...rest, ...byPair.values()]
}

export function filterAlertTopology(topology: GraphResponse): GraphResponse {
  const nodes = topology.nodes.filter(isAlertNode)
  const ids = new Set(nodes.map((n) => n.id))
  const edges = dedupeCausedEdges(
    topology.edges.filter(
      (e) =>
        ids.has(e.from) &&
        ids.has(e.to) &&
        (e.label === "CAUSED" || e.label === "RESULTED_IN"),
    ),
  )
  return {
    ...topology,
    nodes,
    edges,
    notifications: topology.notifications,
  }
}

export function filterAlertAttackTrees(trees: GraphTreeNode[]): GraphTreeNode[] {
  return trees.filter((t) => t.type === "Alert")
}

export function filterAlertAttackTreeResponse(
  response: AttackTreeResponse,
): AttackTreeResponse {
  return {
    ...response,
    attack_trees: filterAlertAttackTrees(response.attack_trees),
  }
}

export function parseAlertTimestamp(props: Record<string, unknown>): number {
  const raw = props.timestamp
  if (typeof raw !== "string") return 0
  const t = Date.parse(raw)
  return Number.isNaN(t) ? 0 : t
}

export function orderAlertChain(
  nodes: GraphNode[],
  edges: GraphEdge[],
): GraphNode[] {
  if (nodes.length <= 1) return nodes
  const outDegree = new Map<string, number>()
  const next = new Map<string, string>()
  for (const n of nodes) outDegree.set(n.id, 0)
  for (const e of edges) {
    if (e.label !== "CAUSED" && e.label !== "RESULTED_IN") continue
    next.set(e.from, e.to)
    outDegree.set(e.to, (outDegree.get(e.to) ?? 0) + 1)
  }
  const roots = nodes.filter((n) => (outDegree.get(n.id) ?? 0) === 0)
  const start =
    roots[0] ??
    [...nodes].sort(
      (a, b) =>
        parseAlertTimestamp(a.properties) - parseAlertTimestamp(b.properties),
    )[0]
  const ordered: GraphNode[] = []
  const seen = new Set<string>()
  let cur: GraphNode | undefined = start
  while (cur && !seen.has(cur.id)) {
    ordered.push(cur)
    seen.add(cur.id)
    const nid = next.get(cur.id)
    cur = nid ? nodes.find((n) => n.id === nid) : undefined
  }
  for (const n of nodes) {
    if (!seen.has(n.id)) ordered.push(n)
  }
  return ordered
}

export type LayoutNode = GraphNode & { x: number; y: number }

const NODE_RADIUS = 22
const MIN_H_GAP = 140
const CANVAS_HEIGHT = 420

export function layoutAlertChain(
  nodes: GraphNode[],
  edges: GraphEdge[],
): { layout: LayoutNode[]; width: number; height: number } {
  const ordered = orderAlertChain(nodes, edges)
  const width = Math.max(640, ordered.length * MIN_H_GAP + 80)
  const y = CANVAS_HEIGHT / 2
  const step = width / (ordered.length + 1)
  const layout = ordered.map((node, i) => ({
    ...node,
    x: step * (i + 1),
    y,
  }))
  return { layout, width, height: CANVAS_HEIGHT }
}

export { NODE_RADIUS, MIN_H_GAP, CANVAS_HEIGHT }

File: correlation/graph_crud/alert_centric.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from graph_core.neo4j_driver import run_read_query
from graph_core.neo4j_sanitize import sanitize_neo4j_value
from graph_crud.findings import get_finding
from graph_schemas.exploration import (
    AttackTreeResponse,
    GraphEdge,
    GraphExplorationResponse,
    GraphNode,
    GraphTreeNode,
    HighlightInfo,
)

_INCIDENT_ALERTS_QUERY = """
MATCH (inc:Incident {incident_id: $incident_id})<-[:PART_OF_INCIDENT]-(a:Alert)
WHERE size($alert_ids) = 0 OR a.alert_row_id IN $alert_ids
RETURN
    elementId(a) AS nid,
    labels(a) AS nlabels,
    properties(a) AS nprops
ORDER BY a.timestamp ASC
"""

_ALERTS_BY_ID_QUERY = """
MATCH (a:Alert)
WHERE a.alert_row_id IN $alert_ids
RETURN
    elementId(a) AS nid,
    labels(a) AS nlabels,
    properties(a) AS nprops
ORDER BY a.timestamp ASC
"""

_ALERTS_SUBGRAPH_QUERY = """
MATCH (a:Alert)
WHERE a.alert_row_id IN $alert_ids
WITH collect(DISTINCT a) AS alerts
UNWIND alerts AS node
WITH collect(DISTINCT node) AS nodes
UNWIND nodes AS node
OPTIONAL MATCH (node)-[r]-(m)
WHERE m IN nodes OR (m:Identity OR m:Asset OR m:IOC)
RETURN
    elementId(node) AS nid,
    labels(node) AS nlabels,
    properties(node) AS nprops,
    elementId(r) AS rid,
    type(r) AS rtype,
    elementId(startNode(r)) AS from_id,
    elementId(endNode(r)) AS to_id,
    properties(r) AS rprops
"""

_INCIDENT_SUBGRAPH_QUERY = """
MATCH (inc:Incident {incident_id: $incident_id})<-[:PART_OF_INCIDENT]-(a:Alert)
WHERE size($alert_ids) = 0 OR a.alert_row_id IN $alert_ids
WITH inc, collect(DISTINCT a) AS alerts
WITH [inc] + alerts AS seedNodes
UNWIND seedNodes AS node
WITH collect(DISTINCT node) AS nodes
UNWIND nodes AS node
OPTIONAL MATCH (node)-[r]-(m)
WHERE m IN nodes AND elementId(node) <= elementId(m)
RETURN
    elementId(node) AS nid,
    labels(node) AS nlabels,
    properties(node) AS nprops,
    elementId(r) AS rid,
    type(r) AS rtype,
    elementId(startNode(r)) AS from_id,
    elementId(endNode(r)) AS to_id,
    properties(r) AS rprops
"""


def _label_for_node(labels: list[str], props: dict[str, Any]) -> str:
    if "Alert" in labels:
        return str(props.get("name") or props.get("alert_row_id") or "Alert")
    if "Asset" in labels:
        return str(props.get("name") or props.get("primary_identifier") or "Asset")
    if "Identity" in labels:
        return str(props.get("name") or props.get("primary_identifier") or "Identity")
    if "IOC" in labels:
        return str(props.get("value") or props.get("primary_identifier") or "IOC")
    if "Incident" in labels:
        return str(props.get("title") or props.get("incident_id") or "Incident")
    return str(props.get("name") or props.get("primary_identifier") or "Node")


def _merge_subgraph_rows(
    caused_nodes: list[GraphNode],
    caused_edges: list[GraphEdge],
    subgraph_rows: list[dict[str, Any]],
) -> tuple[list[GraphNode], list[GraphEdge]]:
    nodes_by_id = {n.id: n for n in caused_nodes}
    edges_by_id = {e.id: e for e in caused_edges}

    for row in subgraph_rows:
        nid = row.get("nid")
        if nid is not None:
            sid = str(nid)
            if sid not in nodes_by_id:
                labels = row.get("nlabels") or []
                props = sanitize_neo4j_value(row.get("nprops") or {})
                group = list(labels) if labels else ["Node"]
                nodes_by_id[sid] = GraphNode(
                    id=sid,
                    label=_label_for_node(group, props),
                    group=group,
                    properties=props,
                )
        rid = row.get("rid")
        if rid is not None:
            eid = str(rid)
            if eid not in edges_by_id:
                edges_by_id[eid] = GraphEdge(
                    id=eid,
                    **{"from": str(row["from_id"])},
                    to=str(row["to_id"]),
                    label=str(row.get("rtype") or "RELATED"),
                    properties=sanitize_neo4j_value(row.get("rprops") or {}),
                )

    return list(nodes_by_id.values()), list(edges_by_id.values())


def _dedupe_parallel_caused_edges(edges: list[GraphEdge]) -> list[GraphEdge]:
    """One CAUSED edge per alert pair (synthetic seq + Neo4j merge overlap)."""
    by_pair: dict[tuple[str, str], GraphEdge] = {}
    rest: list[GraphEdge] = []
    for edge in edges:
        if edge.label != "CAUSED":
            rest.append(edge)
            continue
        key = (edge.from_, edge.to)
        existing = by_pair.get(key)
        if existing is None:
            by_pair[key] = edge
            continue
        prev_props = dict(existing.properties or {})
        next_props = dict(edge.properties or {})
        narrative = next_props.get("narrative") or prev_props.get("narrative")
        merged = {**prev_props, **next_props}
        if narrative:
            merged["narrative"] = narrative
        if merged.get("time_delta_seconds") is None:
            merged["time_delta_seconds"] = prev_props.get("time_delta_seconds") or next_props.get(
                "time_delta_seconds"
            )
        by_pair[key] = GraphEdge(
            id=existing.id,
            **{"from": existing.from_},
            to=existing.to,
            label="CAUSED",
            properties=merged,
        )
    return rest + list(by_pair.values())


def _parse_timestamp(props: dict[str, Any]) -> datetime:
    raw = props.get("timestamp")
    if raw is None:
        return datetime.min
    if isinstance(raw, datetime):
        return raw
    if isinstance(raw, str):
        try:
            return datetime.fromisoformat(raw.replace("Z", "+00:00"))
        except ValueError:
            return datetime.min
    return datetime.min


def _alert_label(props: dict[str, Any]) -> str:
    name = props.get("name") or props.get("alert_row_id") or "Alert"
    risk = props.get("risk_score")
    if risk is not None:
        return f"Alert: {name} ({risk})"
    return f"Alert: {name}"


def _time_delta_seconds(prev: dict[str, Any], curr: dict[str, Any]) -> Optional[int]:
    t0 = _parse_timestamp(prev)
    t1 = _parse_timestamp(curr)
    if t0 == datetime.min or t1 == datetime.min:
        return None
    return max(0, int((t1 - t0).total_seconds()))


def _bridge_text(delta: Optional[int]) -> str:
    if delta is None:
        return "Correlated Sequence"
    if delta < 60:
        return f"Sequential Step (+{delta}s)"
    minutes = delta // 60
    if minutes < 60:
        return f"Sequential Step (+{minutes}m)"
    hours = minutes // 60
    if hours < 48:
        remainder_m = minutes % 60
        if remainder_m:
            return f"Sequential Step (+{hours}h {remainder_m}m)"
        return f"Sequential Step (+{hours}h)"
    days = hours // 24
    remainder_h = hours % 24
    if remainder_h:
        return f"Sequential Step (+{days}d {remainder_h}h)"
    return f"Sequential Step (+{days}d)"


def _build_caused_chain(
    alert_rows: list[dict[str, Any]],
) -> tuple[list[GraphNode], list[GraphEdge]]:
    nodes: list[GraphNode] = []
    edges: list[GraphEdge] = []
    for row in alert_rows:
        labels = row.get("nlabels") or []
        if "Alert" not in labels:
            continue
        props = sanitize_neo4j_value(row.get("nprops") or {})
        group = list(labels) if labels else ["Alert"]
        nid = str(row["nid"])
        nodes.append(
            GraphNode(
                id=nid,
                label=str(props.get("name") or props.get("alert_row_id") or "Alert"),
                group=group,
                properties=props,
            )
        )
    for i in range(len(nodes) - 1):
        prev_props = nodes[i].properties
        curr_props = nodes[i + 1].properties
        delta = _time_delta_seconds(prev_props, curr_props)
        edges.append(
            GraphEdge(
                id=f"seq_{nodes[i].id}_{nodes[i + 1].id}",
                **{"from": nodes[i].id},
                to=nodes[i + 1].id,
                label="CAUSED",
                properties={
                    "time_delta_seconds": delta,
                    "confidence": "chronological_sequence",
                    "narrative": _bridge_text(delta),
                },
            )
        )
    return nodes, edges


def build_alert_centric_tree(
    nodes: list[GraphNode],
    _edges: list[GraphEdge],
) -> list[GraphTreeNode]:
    """Flat chronological attack_trees from alert-only CAUSED chain."""
    alert_nodes = [n for n in nodes if "Alert" in n.group]
    alert_nodes.sort(key=lambda n: _parse_timestamp(n.properties))

    trees: list[GraphTreeNode] = []
    for i, node in enumerate(alert_nodes, start=1):
        props = node.properties
        edge_context: Optional[str] = None
        if i > 1:
            delta = _time_delta_seconds(alert_nodes[i - 2].properties, props)
            edge_context = _bridge_text(delta)

        trees.append(
            GraphTreeNode(
                step=str(i),
                node_id=node.id,
                name=_alert_label(props),
                type="Alert",
                timestamp=str(props.get("timestamp") or ""),
                risk_score=int(props["risk_score"])
                if props.get("risk_score") is not None
                else None,
                edge_context=edge_context,
                children=[],
                expandable=False,
            )
        )
    return trees


async def build_alert_centric_topology(
    finding_id: str,
) -> GraphExplorationResponse | None:
    finding = await get_finding(finding_id)
    if finding is None:
        return None

    details = finding.details or {}
    incident_id = details.get("incident_id")
    if not incident_id:
        return GraphExplorationResponse(
            nodes=[],
            edges=[],
            highlight_info=HighlightInfo(),
            message=(
                "No incident_id on finding; alert-centric view requires an incident-bound finding."
            ),
        )

    alert_ids: list[str] = []
    for alert in details.get("contributing_alerts") or []:
        aid = alert.get("alert_row_id")
        if aid:
            alert_ids.append(str(aid))

    rows = await run_read_query(
        _INCIDENT_ALERTS_QUERY,
        {"incident_id": str(incident_id), "alert_ids": alert_ids},
    )
    used_fallback = False
    if not rows and alert_ids:
        rows = await run_read_query(_ALERTS_BY_ID_QUERY, {"alert_ids": alert_ids})
        used_fallback = bool(rows)

    if not rows:
        return GraphExplorationResponse(
            nodes=[],
            edges=[],
            highlight_info=HighlightInfo(),
            message=(
                f"No alerts linked to incident {incident_id} in Neo4j. "
                "Seed PART_OF_INCIDENT edges for demo data."
            ),
        )

    caused_nodes, caused_edges = _build_caused_chain(rows)
    if used_fallback:
        subgraph_rows = await run_read_query(
            _ALERTS_SUBGRAPH_QUERY,
            {"alert_ids": alert_ids},
        )
    else:
        subgraph_rows = await run_read_query(
            _INCIDENT_SUBGRAPH_QUERY,
            {"incident_id": str(incident_id), "alert_ids": alert_ids},
        )
    nodes, edges = _merge_subgraph_rows(caused_nodes, caused_edges, subgraph_rows)
    edges = _dedupe_parallel_caused_edges(edges)
    notifications: list[str] = []
    if used_fallback:
        notifications.append(
            "Loaded alerts by contributing_alerts (incident not yet linked in Neo4j)."
        )
    if alert_ids and len([n for n in nodes if "Alert" in n.group]) < len(alert_ids):
        notifications.append(
            "Some contributing_alerts were not found on the incident timeline in Neo4j."
        )

    return GraphExplorationResponse(
        nodes=nodes,
        edges=edges,
        highlight_info=HighlightInfo(),
        message="Success.",
        notifications=notifications or None,
    )


async def build_alert_centric_attack_tree(
    finding_id: str,
) -> AttackTreeResponse | None:
    topo = await build_alert_centric_topology(finding_id)
    if topo is None:
        return None
    if not topo.nodes:
        return AttackTreeResponse(
            attack_trees=[],
            message=topo.message or "No attack trees for this finding.",
            notifications=topo.notifications,
        )
    trees = build_alert_centric_tree(
        [n for n in topo.nodes if "Alert" in n.group],
        [e for e in topo.edges if e.label == "CAUSED"],
    )
    return AttackTreeResponse(
        attack_trees=trees,
        message="Success.",
        notifications=topo.notifications,
    )
